### PR TITLE
Support work.result() to get result tensors for allreduce for Gloo, NCCL backends (resubmission of  #43386)

### PR DESCRIPTION
**Pull request by mrzzd**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43970 Support work.result() to get result tensors for allreduce for Gloo, NCCL backends (resubmission of  #43386)**

Original commit changeset: 27fbeb161706

Differential Revision: [D23455047](https://our.internmc.facebook.com/intern/diff/D23455047/)

### CODE REVIEW

**Comment by pritamdamania87** [comment]
We can close out this PR since the original diff has been reverted?

File: .circleci/config.yml [hunk 1]
```diff
@@ -6315,12 +6315,6 @@ workflows:
           name: pytorch_linux_xenial_cuda10_1_cudnn7_py3_multigpu_test
           requires:
             - pytorch_linux_xenial_cuda10_1_cudnn7_py3_gcc7_build
-          filters:
-            branches:
-              only:
-                - master
-                - /ci-all\/.*/
-                - /release\/.*/
```
Comment (pritamdamania87) [inline .circleci/config.yml:6323]: What are the reason for these changes in the ci config?

File: test/distributed/test_distributed.py [hunk 2]
```diff
@@ -857,6 +857,45 @@ def test_reduce_full_group_max(self):
...
+                    work.result()
+            except AssertionError:
+                # Exception was not raised, ensure is_completed()
+                self.assertTrue(work.is_completed())
+
+            work.wait()
+
+            result = work.result()
+            expected_value = 2 + (10 * (len(group) - 1))
+            self.assertEqual(result, [_build_tensor(src + 1, expected_value)])
```
Comment (pritamdamania87) [inline test/distributed/test_distributed.py:896]: nit: Since allreduce is in place we can just check if result is equal to tensor?

File: torch/lib/c10d/ProcessGroupGloo.cpp [hunk 3]
```diff
@@ -810,6 +811,15 @@ class AsyncAllreduceWork : public ProcessGroupGloo::AsyncWork {
     GENERATE_ALL_TYPES(dtype, getFunction, fn, op);
     return fn;
   }
+
+
+  std::vector<at::Tensor> result() override {
+    TORCH_CHECK(isCompleted());
```
Comment (pritamdamania87) [inline torch/lib/c10d/ProcessGroupGloo.cpp:817]: Add an error message to `TORCH_CHECK` here mentioning that work needs to be completed before calling `result()`.

File: torch/lib/c10d/ProcessGroupGloo.cpp [hunk 4]
```diff
@@ -810,6 +811,15 @@ class AsyncAllreduceWork : public ProcessGroupGloo::AsyncWork {
...
+    return outputs_;
+  }
+
+ protected:
+  std::vector<at::Tensor> outputs_;
 };
 
 class AsyncAllreduceCoalescedWork : public AsyncAllreduceWork {
```
Comment (pritamdamania87) [inline torch/lib/c10d/ProcessGroupGloo.cpp:825]: Apart from `allreduce`, there are other ops like barrier, scatter, gather, reduce etc. can we add this functionality to those as well? Can be in a separate PR on top of this.

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 5]
```diff
@@ -246,7 +246,8 @@ ProcessGroupNCCL::WorkNCCL::~WorkNCCL() {}
 
 bool ProcessGroupNCCL::WorkNCCL::isCompleted() {
   checkAndSetException();
-  return exception() || finishedGPUExecutionInternal();
+  bool result = exception() || finishedGPUExecutionInternal();
+  return result;
```
Comment (pritamdamania87) [inline torch/lib/c10d/ProcessGroupNCCL.cpp:250]: Why not have this in a single line as before?

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 6]
```diff
@@ -782,6 +783,11 @@ std::shared_ptr<ProcessGroupNCCL::WorkNCCL> ProcessGroupNCCL::initWork(
   return std::make_shared<ProcessGroupNCCL::WorkNCCL>(devices);
 }
 
+std::vector<at::Tensor> ProcessGroupNCCL::WorkNCCL::result() {
+  TORCH_CHECK(isCompleted());
```
Comment (pritamdamania87) [inline torch/lib/c10d/ProcessGroupNCCL.cpp:787]: This is not necessarily true for NCCL. `isCompleted()` returns true only when the work has completed successfully on the GPU. But we can still retrieve the `result` for the Work and perform async processing. I think we shouldn't have this check for NCCL.

**Review by pritamdamania87 — CHANGES_REQUESTED** [review]
If this is a resubmission of https://github.com/pytorch/pytorch/pull/43386, can we update the PR title and summary to reflect this appropriately?

File: .circleci/config.yml (hunk 1 above)
Comment (mrzzd) [reply .circleci/config.yml:6323]: This is temporary change to force tests run for this PR. I'll undo it in next update to PR

File: test/distributed/test_distributed.py (hunk 2 above)
Comment (mrzzd) [reply test/distributed/test_distributed.py:896]: > nit: Since allreduce is in place we can just check if result is equal to tensor?

Being equal to tensor does not mean result is ready, it could be both of them have non-finalized values. I see that's being tested in other unitttests, but still, better to not depend on them.

File: torch/lib/c10d/ProcessGroupNCCL.cpp (hunk 6 above)
Comment (mrzzd) [reply torch/lib/c10d/ProcessGroupNCCL.cpp]: Good point, removed.

File: torch/lib/c10d/ProcessGroupGloo.cpp (hunk 4 above)
Comment (mrzzd) [reply torch/lib/c10d/ProcessGroupGloo.cpp:825]: Let's do it in separate PR

**Comment by mrzzd** [comment]
Verified that failing test also passed: https://app.circleci.com/pipelines/github/pytorch/pytorch/210229/workflows/86bde47b-f2da-48e3-a618-566ae2713102/jobs/7253683

File: test/distributed/test_distributed.py (hunk 2 above)
Comment (pritamdamania87) [reply test/distributed/test_distributed.py:913]: We are calling `work.wait()` before the check though right? That should synchronize the appropriate streams to ensure this is safe to do.

File: torch/lib/c10d/ProcessGroupGloo.cpp [hunk 7]
```diff
@@ -814,7 +814,7 @@ class AsyncAllreduceWork : public ProcessGroupGloo::AsyncWork {
 
 
   std::vector<at::Tensor> result() override {
-    TORCH_CHECK(isCompleted());
+    TORCH_CHECK(isCompleted(), "Work needs to be completed before calling result().");
```
Comment (pritamdamania87) [inline torch/lib/c10d/ProcessGroupGloo.cpp:817]: nit: Extend this to mention, you should call `wait()` before `result()`.

**Review by pritamdamania87 — APPROVED** [review]
Looks good to me! Can we also update the PR summary appropriate? It currently still mentions "#43970 Back out "Revert D23216393: Support work.result() to get result tensors for allreduce for Gloo, NCCL backends"". Also, please check all the builds are clean before landing.

**Comment by mrzzd** [comment]
> We can close out this PR since the original diff has been reverted?

I just rebased with latest changes in trunk. I am going to submit this as soon as testing finishes.
